File: python/mirage/commands/builtin/generic/mktemp.py
import random
import string
from collections.abc import Awaitable, Callable, Mapping
from dataclasses import dataclass

from mirage.commands.config import CommandOpts
from mirage.commands.spec import SPECS
from mirage.commands.spec.flag_view import FlagView
from mirage.commands.spec.types import CommandName, FlagValue
from mirage.commands.spec.usage import extra_operand_error
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec
from mirage.utils.key_prefix import rekey

_ALPHABET = string.ascii_letters + string.digits


def _rand_suffix(length: int) -> str:
    return "".join(random.choices(_ALPHABET, k=length))


def _build_path(p: str | PathSpec | None, t: bool, texts: tuple[str, ...],
                suffix: str) -> tuple[PathSpec, PathSpec]:
    template = texts[0] if texts else "tmp.XXXXXXXXXX"
    if t:
        parent = PathSpec.from_str_path("/tmp")
    elif p is not None:
        parent = p if isinstance(p, PathSpec) else PathSpec.from_str_path(p)
    elif "/" in template:
        # An explicit path template names its own directory (GNU); only a
        # bare template with no -p/-t falls back to the temp dir.
        head, _, template = template.rpartition("/")
        parent = PathSpec.from_str_path(head or "/")
    else:
        parent = PathSpec.from_str_path("/tmp")
    i = len(template)
    while i > 0 and template[i - 1] == "X":
        i -= 1
    if i < len(template):
        name = template[:i] + _rand_suffix(len(template) - i) + suffix
    else:
        name = f"{template}.{_rand_suffix(8)}"
    virtual = f"{parent.virtual.rstrip('/')}/{name}"
    return PathSpec.from_str_path(
        virtual,
        rekey(parent.virtual, parent.vfs_path, virtual),
    ), parent


async def mktemp(
    *texts: str,
    mkdir_fn: Callable[..., Awaitable[None]],
    write_bytes_fn: Callable[..., Awaitable[None]],
    d: bool = False,
    p: str | PathSpec | None = None,
    t: bool = False,
    dry_run: bool = False,
    suffix: str = "",
    quiet: bool = False,
) -> tuple[ByteSource | None, IOResult]:
    if len(texts) > 1:
        raise extra_operand_error(CommandName.MKTEMP, texts[1])
    path, parent = _build_path(p, t, texts, suffix)
    if not dry_run:
        # -q suppresses diagnostics about file/directory creation only
        # (GNU); usage errors and internal failures still propagate.
        try:
            await mkdir_fn(parent, parents=True)
            if d:
                await mkdir_fn(path)
            else:
                await write_bytes_fn(path, b"")
        except OSError:
            if not quiet:
                raise
            return None, IOResult(exit_code=1)
    return (path.virtual + "\n").encode(), IOResult()


__all__ = ["mktemp"]


@dataclass(frozen=True, slots=True)
class MktempFlags:
    directory: bool = False
    tmpdir: "str | PathSpec | None" = None
    template_mode: bool = False
    dry_run: bool = False
    suffix: str = ""
    quiet: bool = False


def parse_flags(flags: Mapping[str, FlagValue]) -> MktempFlags:
    fl = FlagView(flags, spec=SPECS["mktemp"])
    tmpdir_flag = fl.raw("tmpdir")
    tmpdir: str | PathSpec | None
    if isinstance(tmpdir_flag, (str, PathSpec)):
        tmpdir = tmpdir_flag
    elif tmpdir_flag is True:
        tmpdir = "/tmp"
    else:
        p_flag = fl.raw("p")
        tmpdir = p_flag if isinstance(p_flag, (str, PathSpec)) else None
    return MktempFlags(
        directory=fl.as_bool("directory"),
        tmpdir=tmpdir,
        template_mode=fl.as_bool("t"),
        dry_run=fl.as_bool("dry_run"),
        suffix=fl.as_str("suffix") or "",
        quiet=fl.as_bool("quiet"),
    )


async def mktemp_generic(
    paths: list[PathSpec],
    texts: list[str],
    opts: CommandOpts,
    mkdir_fn: Callable[..., Awaitable[None]],
    write_bytes_fn: Callable[..., Awaitable[None]],
) -> tuple[ByteSource | None, IOResult]:
    parsed = parse_flags(opts.flags)
    return await mktemp(*texts,
                        mkdir_fn=mkdir_fn,
                        write_bytes_fn=write_bytes_fn,
                        d=parsed.directory,
                        p=parsed.tmpdir,
                        t=parsed.template_mode,
                        dry_run=parsed.dry_run,
                        suffix=parsed.suffix,
                        quiet=parsed.quiet)

File: typescript/packages/core/src/commands/builtin/generic/mktemp.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { specOf } from '../../spec/builtins.ts'
import { FlagView } from '../../spec/flag_view.ts'
import { IOResult, type ByteSource } from '../../../io/types.ts'
import { PathSpec } from '../../../types.ts'
import type { CommandFnResult, CommandOpts } from '../../config.ts'
import { rstripSlash } from '../../../utils/slash.ts'
import { mountKey } from '../../../utils/key_prefix.ts'
import { extraOperandError } from '../../spec/usage.ts'
import { CommandName } from '../../spec/types.ts'

const ENC = new TextEncoder()

function randomSuffix(length: number): string {
  const chars = 'abcdefghijklmnopqrstuvwxyz0123456789'
  let out = ''
  for (let i = 0; i < length; i++) {
    out += chars[Math.floor(Math.random() * chars.length)] ?? ''
  }
  return out
}

function makePathSpec(virtual: string, mountPrefix: string): PathSpec {
  return new PathSpec({
    virtual,
    directory: virtual,
    vfsPath: mountKey(virtual, mountPrefix),
    resolved: true,
  })
}

export async function mktempGeneric(
  texts: string[],
  opts: CommandOpts,
  mkdir: (p: PathSpec, parents?: boolean) => Promise<void>,
  write: (p: PathSpec, data: Uint8Array) => Promise<void>,
): Promise<CommandFnResult> {
  const fl = new FlagView(opts.flags, specOf('mktemp'))
  if (texts.length > 1) throw extraOperandError(CommandName.MKTEMP, texts[1] ?? '')
  const tFlag = fl.asBool('t')
  const directory = fl.asBool('directory')
  const dryRun = fl.asBool('dry_run')
  const suffix = fl.asStr('suffix') ?? ''
  const tmpdirValue: unknown = fl.raw('p') ?? fl.raw('tmpdir')
  const templateArg = texts[0]
  let template = templateArg !== undefined && templateArg !== '' ? templateArg : 'tmp.XXXXXXXXXX'
  let parent: string
  if (tFlag) {
    parent = '/tmp'
  } else if (tmpdirValue instanceof PathSpec) {
    parent = tmpdirValue.virtual
  } else if (typeof tmpdirValue === 'string') {
    parent = tmpdirValue
  } else if (template.includes('/')) {
    // An explicit path template names its own directory (GNU); only a bare
    // template with no -p/-t falls back to the temp dir.
    const idx = template.lastIndexOf('/')
    parent = template.slice(0, idx) || '/'
    template = template.slice(idx + 1)
  } else {
    parent = '/tmp'
  }
  let xCount = 0
  while (xCount < template.length && template.charCodeAt(template.length - 1 - xCount) === 88) {
    xCount += 1
  }
  let name: string
  if (xCount > 0) {
    name = template.slice(0, template.length - xCount) + randomSuffix(xCount) + suffix
  } else {
    name = `${template}.${randomSuffix(8)}`
  }
  const path = `${rstripSlash(parent)}/${name}`
  if (!dryRun) {
    const mountPrefix = opts.mountPrefix ?? ''
    const quiet = fl.asBool('quiet')
    try {
      await mkdir(makePathSpec(parent, mountPrefix), true)
      if (directory) {
        await mkdir(makePathSpec(path, mountPrefix))
      } else {
        await write(makePathSpec(path, mountPrefix), new Uint8Array(0))
      }
    } catch (error) {
      // -q suppresses diagnostics about file/directory creation only
      // (GNU); usage errors and internal failures still propagate.
      if (!quiet) throw error
      return [null, new IOResult({ exitCode: 1 })]
    }
  }
  const result: ByteSource = ENC.encode(path + '\n')
  return [result, new IOResult()]
}
